truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -354,126 +545,144 @@ public Object addSpecial(RubyBasicObject a, RubyBasicObject b) {
             else
                 return b;
         }
-
-        // TODO (pitr 28-may-2015): should it support coerce?
     }
 
-    @CoreMethod(names = "add", required = 2)
-    public abstract static class AddNode extends BigDecimalCoreMethodNode {
+    @CoreMethod(names = {"+"}, required = 1)
+    public abstract static class AddOpNode extends AbstractAddNode {
 
-        public AddNode(RubyContext context, SourceSection sourceSection) {
+        public AddOpNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
         }
 
-        @TruffleBoundary
-        private Object addBigDecimal(RubyBasicObject a, BigDecimal b, int precision) {
-            return createRubyBigDecimal(getBigDecimalValue(a).add(b, new MathContext(precision)));
+        @Specialization(guards = {
+                "isNormal(a)",
+                "isRubyBigDecimal(b)",
+                "isNormal(b)"})
```

Comment:
It may not matter much here, but the predicates are evaluated in order and short-circuit.  Assuming `isNormal` has uniform cost, both calls are either faster than or slower than the `isRubyBigDecimal` check and should probably be ordered as such.